pkg/covermerger/covermerger.go

```diff
@@ -123,38 +146,81 @@ func makeRecord(fields, schema []string) FileRecord {
 	return record
 }
 
+const (
+	BaseManual = iota
+	BaseLastUpdated
+)
+
 type Config struct {
-	Workdir       string
-	skipRepoClone bool
+	Workdir             string
+	skipRepoClone       bool
+	BaseType            int              // BaseManual, BaseLastUpdated.
+	Base                RepoBranchCommit // used by BaseManual
+	getFileVersionsMock func(*Config, string, []RepoBranchCommit) (fileVersions, error)
 }
 
-func AggregateStreamData(c *Config, stream io.Reader, base RepoBranchCommit,
-) (map[string]*MergeResult, error) {
-	stat := make(map[string]*MergeResult)
+func isSchema(fields, schema []string) bool {
+	if len(fields) != len(schema) {
+		return false
+	}
+	for i := 0; i < len(fields); i++ {
+		if fields[i] != schema[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func MergeCSVData(config *Config, reader io.Reader) (map[string]*MergeResult, error) {
 	var schema []string
-	targetFile := ""
-	var records FileRecords
-	csvReader := csv.NewReader(stream)
+	csvReader := csv.NewReader(reader)
 	if fields, err := csvReader.Read(); err != nil {
 		return nil, fmt.Errorf("failed to read schema: %w", err)
 	} else {
 		schema = fields
 	}
-	for {
-		fields, err := csvReader.Read()
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			return nil, fmt.Errorf("failed to read CSV line: %w", err)
+	errStdinReadChan := make(chan error, 1)
+	recordsChan := make(chan FileRecord)
+	go func() {
+		defer close(recordsChan)
+		for {
+			fields, err := csvReader.Read()
+			if err == io.EOF {
+				break
+			}
+			if err != nil {
+				errStdinReadChan <- fmt.Errorf("failed to read CSV line: %w", err)
+				return
+			}
+			if isSchema(fields, schema) {
+				// The input may be the merged CVS files with multiple schemas.
+				continue
+			}
+			recordsChan <- makeRecord(fields, schema)
 		}
-		record := makeRecord(fields, schema)
-		curTargetFile := record[keyFilePath]
+		errStdinReadChan <- nil
+	}()
+	mergeResult, errMerging := MergeChanData(config, recordsChan)
+	errStdinRead := <-errStdinReadChan
+	if errMerging != nil || errStdinRead != nil {
+		return nil, fmt.Errorf("errors merging stdin data:\nmerger err: %w\nstdin reader err: %w",
+			errMerging, errStdinRead)
+	}
+	return mergeResult, nil
+}
+
+func MergeChanData(c *Config, recordsChan <-chan FileRecord) (map[string]*MergeResult, error) {
```

Comment:
In the code below, you assume that, for every file name, there's a consecutive batch of the coverage of that file in different repositories.

If you just contatenate the CSV files and feed them into the tool, what are those CSV files? If they are per-manager, the same file will be mentioned in all of them, and here you just overwrite the `stat` map's value each time you encounter that file name.

Reply:
The data is selected by:
  SELECT
    kernel_repo, kernel_branch, kernel_commit, file_path, sl, SUM(hit_count) as hit_count
  FROM syzkaller.syzbot_coverage.'$namespace'
  WHERE
    TIMESTAMP_TRUNC(timestamp, DAY) >= "'$from_date'" AND
    TIMESTAMP_TRUNC(timestamp, DAY) <= "'$to_date'" AND
    version = 1
  GROUP BY file_path, kernel_commit, kernel_repo, kernel_branch, sl
  ORDER BY file_path

It is grouped by file_path.
To address your concern there are two options:
1. Read the whole stream to some map (map[file_name]signals) and process the map items. It is doable if the tool will be used to process to process relatively small coverage histories.
3. Detect the fact input is not sorted.

The biggest aggregation I can imagine is something like 10 years long coverage history from all namespaces.
100k functions * 6 callbacks/function * 15 namespaces * 10 commits_daily_in_every_namespace * 3650 days =~ 300G csv lines.
Let's expect the input to be sorted. I'll add the detection.

Reply:
Still, what are the CSV files now? Are they per syz-manager instance or they are just that huge sorted bigquery output that was just split into multiple files because of its size?

Reply:
> The biggest aggregation I can imagine is something like 10 years long coverage history from all namespaces.

Will we ever need one single aggregation over all the data of 10 years? I guess, we'd rather be more interested to see it e.g. in the form of a sliding window of some reasonable size (e.g. a week or month). And, in that case, the code will only be run on each sliding window's worth of data, which is way less than the calculated 300G.

I'd also assume that most of that data is repetitive (e.g. we have a very limited set of possible values for repo/branch/filename, but these strings account for a large %% of the CSV file size). The processed/aggregated representation can be much much more compact.

Reply:
The case I'd like to additionally cover by the order verification is the csv files misorder.

Reply:
done